### ISSUE TITLE
[intake] Implement the full backward pass for the GoogLeNet CIFAR-10 trai

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5518]
## Intake

**Idea:** Implement the full backward pass for the GoogLeNet CIFAR-10 training script per ProjectOdyssey issue #3184, then validate by training one epoch with decreasing loss.

**Context:** Existing GitHub issue #3184 covers this work; scope strictly to it. The mesh should decompose into at most 3 dependency-ordered task-agent tasks, and the final PR must contain 'Closes #3184'.

Interview transcript and researched brief follow as comments.

**Comment by mvillmow** [comment]
**Q:** The issue's own convergence proof is an in-file `--self-test` (3-step loss decrease on synthetic input) because CIFAR-10 may be absent, yet your task statement requires validating by 'training one epoch with decreasing loss' on CIFAR-10 — which is authoritative for acceptance: the unconditional synthetic self-test, a real 1-epoch CIFAR-10 run (and if so, is the dataset available in the mesh environment or must one epoch be gated/skipped when absent), or both?
**A:** Operator brief: scope strictly to ProjectOdyssey issue #3184 (full backward pass for the GoogLeNet CIFAR-10 training script, examples/googlenet-cifar10/train.mojo). If #3184 already has a plan with state:plan-go, follow it. Deliverables per the issue: forward pass with activation caching through all 9 Inception modules, full backward propagation through all 9 Inception modules, momentum velocity tensors, SGD+momentum parameter updates. Validation: run one training epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing loss. Keep the workflow to at most 3 tasks assigned to pipeline.task-agent, dependency-ordered. The final PR must include 'Closes #3184'. _(via console)_

**Comment by mvillmow** [comment]
**Q:** The plan is a single ~3200-line edit to `train.mojo` plus a small test append; how should this be split across the ≤3 dependency-ordered task-agent tasks — e.g. (1) forward-pass+velocities scaffolding, (2) backward-pass+SGD updates, (3) tests+validation — and must each task compile independently, or is only the final task required to build?
**A:** Operator brief: scope strictly to ProjectOdyssey issue #3184 (full backward pass for the GoogLeNet CIFAR-10 training script, examples/googlenet-cifar10/train.mojo). If #3184 already has a plan with state:plan-go, follow it. Deliverables per the issue: forward pass with activation caching through all 9 Inception modules, full backward propagation through all 9 Inception modules, momentum velocity tensors, SGD+momentum parameter updates. Validation: run one training epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing loss. Keep the workflow to at most 3 tasks assigned to pipeline.task-agent, dependency-ordered. The final PR must include 'Closes #3184'. _(via console)_

**Comment by mvillmow** [comment]
**Q:** Should the work land as one PR containing 'Closes #3184' (matching 'one PR per issue'), with the ≤3 agent tasks all committing to a single shared feature branch, or is a different branch/PR topology intended?
**A:** Operator brief: scope strictly to ProjectOdyssey issue #3184 (full backward pass for the GoogLeNet CIFAR-10 training script, examples/googlenet-cifar10/train.mojo). If #3184 already has a plan with state:plan-go, follow it. Deliverables per the issue: forward pass with activation caching through all 9 Inception modules, full backward propagation through all 9 Inception modules, momentum velocity tensors, SGD+momentum parameter updates. Validation: run one training epoch on CIFAR-10 (a bounded subset is acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing loss. Keep the workflow to at most 3 tasks assigned to pipeline.task-agent, dependency-ordered. The final PR must include 'Closes #3184'. _(via console)_

**Comment by mvillmow** [comment]
## Researched brief

I have confirmed all the key facts. Both VGG-16 (`compute_gradients` at line 41) and ResNet-18 (`train_epoch` with `velocities` at line 164) are working prior art for the exact pattern. The GoogLeNet `train.mojo` is a documented placeholder. The plan is approved (`state:plan-go`). I have everything needed to write the brief.

---

# Researched Brief: Full Backward Pass for GoogLeNet CIFAR-10 (#3184)

## Feasibility — High

This is a **mechanical port of an already-established, twice-proven pattern**, not novel research. Every primitive the work needs already exists and is exercised by sibling training scripts:

- **The plan is pre-approved.** Issue #3184 carries `state:plan-go` ("Plan approved by reviewer; implementer may proceed") plus a fully-specified implementation plan in the issue body: exact file/function/signature list, a 222-velocity-tensor count derived from `model.mojo`, per-module `split_with_indices` cumulative-sum tables, and resolved answers to a prior review's five findings. Per the operator brief, the implementer follows this plan.
- **The current state is a documented placeholder.** `examples/googlenet_cifar10/train.mojo:87–110` is a comment block describing the ~3500-line backward pass in prose ("Below is the STRUCTURE of the backward pass"); `train_epoch` does no gradient work and `main()` prints an "IMPLEMENTATION STATUS" banner (line 445). There is no velocity threading yet.
- **All backward primitives are already implemented and exported** from `projectodyssey.core`: `conv2d_backward`, `maxpool2d_backward`, `global_avgpool2d_backward`, `batch_norm2d_backward`, `relu_backward`, `linear_backward`, `dropout_backward`, `cross_entropy_backward`, plus `split_with_indices` (`core/shape.mojo`), `add` (`core/arithmetic.mojo`), and `sgd_momentum_update_inplace` (`training/optimizers/sgd.mojo`). No new library code is required — this stays strictly inside `examples/googlenet_cifar10/`.

**Primary risk** is not algorithmic but **bookkeeping at scale**: ~3200 lines of hand-inlined forward/backward with 222 parameter tensors, where the velocity-allocation order must exactly mirror the 222 `sgd_momentum_update_inplace` call order, and each of the 9 Inception modules needs correct 4-way `split_with_indices` cumulative-sum indices. The plan mitigates this with (a) a runtime `len(velocities) != 222` fail-loud check, (b) a per-module channel-split comment table, and (c) `pixi run mojo build` as a compile gate after each module. The GLIBC-2.32 constraint means CPU validation should run in the container (`just shell`) or CI, not necessarily the host.

## Prior Art in This Repo

Two sibling scripts implement the *identical* inlined-forward + reverse-backward + SGD-momentum pattern the plan ports from:

| Reference | What it proves |
|---|---|
| `examples/vgg16_cifar10/train.mojo:41` (`compute_gradients`) | The canonical inlined forward+backward+update convention. Wraps `sgd_momentum_update_inplace` with `Float64(lr), Float64(momentum)`; uses `dropout_backward` with `Float64` p; uses `GradientTriple.grad_weights`. This is the template. |
| `examples/resnet18_cifar10/train.mojo:164` (`train_epoch` + `velocities: List[AnyTensor]`) | Proves the momentum-velocity threading pattern (allocate once in `main()`, thread through `train_epoch`), which GoogLeNet's placeholder lacks. ResNet-18 threads 84 velocity tensors; GoogLeNet needs 222. |
| `examples/googlenet_cifar10/model.mojo` | The forward model (`InceptionModule`, `GoogLeNet`, `concatenate_depthwise`) is complete; only the training-side backward is missing. |
| `tests/models/test_googlenet_e2e.mojo` | Establishes that tests **redefine** `InceptionModule` locally rather than import from `examples/` — which is why the plan's convergence self-test must be co-located in `train.mojo` (`--self-test`) rather than under `tests/`. |

The dual-type tensor architecture (ADR-012: `AnyTensor` for interfaces, `Tensor[dtype]` inside layers) and the `backward-pass-catalog.md` (29 documented backward functions) confirm every needed backward is available.

## Proposed Scope (strictly #3184)

Land as **one PR** with `Closes #3184`, decomposed into **3 dependency-ordered `task-agent` tasks on a single shared feature branch** (honoring "one PR per issue"). The natural cut points follow the plan's own compile gates:

1. **Task 1 — Forward + velocity scaffolding.** Update imports; add `_append_inception_velocities`, `initialize_velocities` (222-tensor fail-loud), `_flatten_gap`/`_unflatten_gap_grad`; write the `compute_gradients` forward section (initial block → 9 Inception forwards with full activation caching → GAP → flatten → dropout → linear → cross-entropy). **Gate: `pixi run mojo build`.**
2. **Task 2 — Backward + SGD updates** (blocked by Task 1). Write the reverse-order backward (classifier → 5b → … → 3a → initial block) with the per-module split-index table, then the 222 ordered `sgd_momentum_update_inplace` calls; rewrite `train_epoch` and `main()` to thread velocities and drop the placeholder banner. **Gate: `pixi run mojo build`.**
3. **Task 3 — Validate + tests** (blocked by Task 2). Add the in-file `--self-test` (3-step strict loss decrease + `fc_bias[0]` mutation + finiteness) and the `split_with_indices`↔`concatenate_depthwise` round-trip test in `tests/models/test_googlenet_layers.mojo`. Run one CIFAR-10 epoch (bounded subset acceptable if a full epoch exceeds 60 min on CPU) and capture the decreasing-loss log. Format + pre-commit. **Open the PR with `Closes #3184`.**

Each task compiles at its boundary (build is a per-task gate); only Task 3 must additionally run and produce the validation log.

## Ideated Extensions (file as follow-ups, out of scope here)

1. **Auxiliary classifier heads (inception_4a/4d) with weighted loss.** The real GoogLeNet paper adds two auxiliary softmax branches (0.3-weighted) to combat vanishing gradients through the deep stack. The current model omits them. A follow-up would add the aux heads to `model.mojo` and their weighted-loss backward to `train.mojo` — a faithful-reproduction improvement squarely in ProjectOdyssey's "reproduce classic papers" mission.
2. **Extract the inlined backward into a reusable autograd tape.** The placeholder's own comment ("For actual training, consider using automatic differentiation") points at the real DRY win: VGG-16, ResNet-18, and now GoogLeNet each hand-inline ~3000 lines of near-identical forward+backward. A follow-up could prototype a minimal `AnyTensor` autograd tape (leveraging the existing `gradient_types` / backward-pass catalog) so future paper reproductions declare the forward once and get the backward for free.

---

```yaml
apiVersion: telemachy/v1
metadata:
  name: googlenet-backward-pass-3184
  description: Implement and validate the full GoogLeNet CIFAR-10 backward pass per approved plan in issue #3184
agents:
  - name: task-agent
    program: claude-code
teams:
  - name: implementation
    agents: [task-agent]
    tasks:
      - subject: Add GoogLeNet forward-pass caching and 222 momentum velocity tensors
        description: >
          In examples/googlenet_cifar10/train.mojo, following the approved plan in issue #3184
          and the pattern in examples/vgg16_cifar10/train.mojo:41 and
          examples/resnet18_cifar10/train.mojo:164. Update imports (backward primitives from
          projectodyssey.core, split_with_indices from core.shape, add from core.arithmetic,
          sgd_momentum_update_inplace from training.optimizers, concatenate_depthwise from model).
          Add helpers _append_inception_velocities, initialize_velocities (allocating exactly 222
          zero velocity tensors in the order: initial block 4, then 9 Inception modules x 24 = 216,
          then FC 2, with a runtime fail-loud check len != 222), and _flatten_gap /
          _unflatten_gap_grad. Write the compute_gradients FORWARD section only: initial conv+BN+ReLU
          +maxpool, then all 9 Inception modules with full per-branch activation caching, then
          global_avgpool2d, flatten, dropout(Float64(0.4)), linear, cross_entropy. Verify:
          `pixi run mojo build examples/googlenet_cifar10/train.mojo` compiles cleanly.
        assign_to: task-agent
        blocked_by: []
      - subject: Implement GoogLeNet backward pass and 222 SGD-momentum updates
        description: >
          In examples/googlenet_cifar10/train.mojo, extend compute_gradients with the BACKWARD
          section in reverse order: cross_entropy_backward, linear_backward, dropout_backward,
          _unflatten_gap_grad, global_avgpool2d_backward, then Inception modules 5b->3a each using
          split_with_indices with the per-module cumulative-sum channel table from the issue plan
          (e.g. 5b -> [384,768,896], 3a -> [64,192,224]), branch backwards via relu_backward /
          batch_norm2d_backward (index access bwd[0..2]) / conv2d_backward (use GradientTriple.
          grad_weights), combining 4 branch input-gradients with core.arithmetic.add, plus
          maxpool2d_backward between stages and the initial-block backward. Then emit exactly 222
          sgd_momentum_update_inplace calls in the SAME order as initialize_velocities, each wrapping
          Float64(learning_rate), Float64(momentum). Rewrite train_epoch to call compute_gradients
          and thread mut velocities; update main() to allocate velocities once and remove the
          IMPLEMENTATION STATUS banner. Verify: `pixi run mojo build examples/googlenet_cifar10/train.mojo`.
        assign_to: task-agent
        blocked_by: [Add GoogLeNet forward-pass caching and 222 momentum velocity tensors]
      - subject: Validate one-epoch convergence and open PR closing 3184
        description: >
          Add an in-file _run_self_test dispatched by a --self-test CLI flag in
          examples/googlenet_cifar10/train.mojo that runs compute_gradients for 3 steps on
          deterministic synthetic input and asserts strict loss decrease at each step, fc_bias[0]
          mutation, and loss finiteness. Append test_split_with_indices_inverse_of_
          concatenate_depthwise to tests/models/test_googlenet_layers.mojo and register it in
          that file's main(). Verify unconditionally with
          `pixi run mojo run examples/googlenet_cifar10/train.mojo -- --self-test` and
          `pixi run mojo run tests/models/test_googlenet_layers.mojo` (use the container / `just shell`
          if the host GLIBC is below 2.32). Then run one CIFAR-10 training epoch (a bounded subset is
          acceptable if a full epoch exceeds 60 minutes on CPU) and capture the log showing decreasing
          loss. Run `pixi run mojo format` and pre-commit. Open a single PR from the shared feature
          branch whose body contains 'Closes #3184' and enable auto-merge.
        assign_to: task-agent
        blocked_by: [Implement GoogLeNet backward pass and 222 SGD-momentum updates]
```

**Comment by mvillmow** [comment]
Registered epic #5522 (key `homericintelligence-projectodyssey-5522`).

**Comment by mvillmow** [comment]
Target ProjectOdyssey #3184 (Full backward pass for GoogLeNet) is CLOSED; the tracked work (epic #5522: sub-tasks #5519/#5520/#5521) is all CLOSED. Done.
